### PR TITLE
Add websocket transport to transit relay

### PR DESCRIPTION
**Pull request by bryanchriswhite**
Our team has been using wormhole-william as our project's client library in the browser via a WASM wrapper. This PR adds support for websocket transport to the transit relay. The transit relay needs to use a transport that is compatible with browser environments. Websockets seem to be the obvious solution here; foregoing any webrtc-based approaches (which we're not interested in pursuing right now).

### CODE REVIEW

**Comment by piegamesde** [comment]
For compatibility, I'd like to keep implementations in sync with the specification. At the moment, the latest draft spec for this is https://github.com/magic-wormhole/magic-wormhole-protocols/pull/16. I implemented the client part in the Rust client and am rather satisfied, but feel free to propose changes or alternatives that better match what you have in mind. I think everything except the new hints encoding should be rather uncontroversial.

**Comment by Jacalz** [comment]
I'll try to review this as soon as I can. Wasm support in Fyne is nearing completion, so I'll want to try and get wormhole-gui running there as soon as possible.

File: wormhole/wormhole_test.go [hunk 1]
```diff
@@ -736,23 +767,142 @@ func TestWormholeDirectoryTransportSendRecvDirect(t *testing.T) {
...
+				t.Fatalf("Failed to get verifier code c0=%q c1=%q", c0Verifier, c1Verifier)
+			}
+
+			if c0Verifier != c1Verifier {
+				t.Fatalf("Expected verifiers to match but were different")
+			}
+		})
+	}
+}
+
```
Comment (vu3rdd) [inline wormhole/wormhole_test.go:885]: There is a lot of repetition of test code. Can this be abstracted into another function with proper parameterization? May be something to keep in mind for another PR, not this one.

File: wormhole/wormhole_test.go [hunk 2]
```diff
@@ -736,23 +767,142 @@ func TestWormholeDirectoryTransportSendRecvDirect(t *testing.T) {
...
+	proto   string
 	wg      sync.WaitGroup
 	mu      sync.Mutex
 	streams map[string]net.Conn
 }
 
-func newTestRelayServer() *testRelayServer {
-	l, err := net.Listen("tcp", ":0")
+func newTestTCPRelayServer() *testRelayServer {
+	l, err := net.Listen("tcp4", ":0")
```
Comment (vu3rdd) [inline wormhole/wormhole_test.go:897]: Any specific reason why we are listening only on ipv4?

File: wormhole/wormhole_test.go [hunk 3]
```diff
@@ -736,23 +767,142 @@ func TestWormholeDirectoryTransportSendRecvDirect(t *testing.T) {
...
 	if err != nil {
 		panic(err)
 	}
 
 	rs := &testRelayServer{
 		l:       l,
-		addr:    l.Addr().String(),
+		url:     internal.MustNewSimpleURL("tcp:" + l.Addr().String()),
```
Comment (vu3rdd) [inline wormhole/wormhole_test.go:904]: Is this valid, now that we are using `net/url`?

Comment (vu3rdd) [reply wormhole/wormhole_test.go:904]: Also, do we need `internal/url` anymore? I am guessing we don't need it as we use generic urls now?

**Review by vu3rdd — COMMENTED** [review]
Looks great overall. Left some comments.

File: wormhole/file_transport.go [hunk 4]
```diff
@@ -372,27 +399,43 @@ func (t *fileTransport) makeTransitMsg() (*transitMsg, error) {
...
-			Hints: []transitHintsV1Hint{
-				{
-					Type:     "direct-tcp-v1",
-					Priority: 2.0,
-					Hostname: relayHost,
-					Port:     relayPort,
+		if relayType == "direct-tcp-v1" {
+			var port, err = strconv.Atoi(t.relayURL.Port())
+			if err != nil {
+				return nil, fmt.Errorf("invalid port")
```
Comment (Jacalz) [inline wormhole/file_transport.go:416]: I think it would make more sense to use `errors.New` here instead.

File: wormhole/file_transport.go [hunk 5]
```diff
@@ -449,23 +492,44 @@ func (t *fileTransport) listen() error {
...
+			return nil
+		}
+		conn, err = net.Dial("tcp", addr)
+		if err != nil {
+			return err
+		}
+	case "ws", "wss":
+		c, _, err := websocket.Dial(ctx, t.relayURL.String(), nil)
+		if err != nil {
+			return fmt.Errorf("websocket.Dial failed")
```
Comment (Jacalz) [inline wormhole/file_transport.go:528]: Same as my comment above.

File: wormhole/recv.go [hunk 6]
```diff
@@ -157,7 +157,11 @@ func (c *Client) Receive(ctx context.Context, code string) (fr *IncomingMessage,
 	}
 
 	transitKey := deriveTransitKey(clientProto.sharedKey, appID)
-	transport := newFileTransport(transitKey, appID, c.relayAddr())
+	relayUrl, err := c.relayURL()
+	if err != nil {
+		return nil, fmt.Errorf("Invalid relay URL")
```
Comment (Jacalz) [inline wormhole/recv.go:162]: Same as my comment above.

File: wormhole/send.go [hunk 7]
```diff
@@ -296,15 +292,20 @@ func (c *Client) sendFileDirectory(ctx context.Context, offer *offerMsg, r io.Re
 			}
 		}
 
+		var relayUrl, err = c.relayURL()
+		if err != nil {
+			sendErr(fmt.Errorf("Invalid relay URL"))
```
Comment (Jacalz) [inline wormhole/send.go:297]: Same as my comment above.

File: wormhole/wormhole.go [hunk 8]
```diff
@@ -33,10 +33,10 @@ type Client struct {
 	// DefaultRendezvousURL will be used.
 	RendezvousURL string
 
-	// TransitRelayAddress is the host:port address to offer
+	// TransitRelayURL is the proto://host:port address to offer
 	// to use for file transfers where direct connections are unavailable.
-	// If empty, DefaultTransitRelayAddress will be used.
-	TransitRelayAddress string
+	// If empty, DefaultTransitRelayURL will be used.
+	TransitRelayURL string
```
Comment (Jacalz) [inline wormhole/wormhole.go:39]: We need to be careful given that this is a braking change. I don't know what plans of @psanford are with this but, if we were to release this in a minor release, would it be possible to do this with just deprecating the old field?

File: wormhole/wormhole.go [hunk 9]
```diff
@@ -63,8 +63,8 @@ var (
 	// DefaultRendezvousURL is the default Rendezvous server to use.
 	DefaultRendezvousURL = "ws://relay.magic-wormhole.io:4000/v1"
 
-	// DefaultTransitRelayAddress is the default transit server to ues.
-	DefaultTransitRelayAddress = "transit.magic-wormhole.io:4001"
+	// DefaultTransitRelayURL is the default transit server to ues.
+	DefaultTransitRelayURL = "tcp://transit.magic-wormhole.io:4001"
```
Comment (Jacalz) [inline wormhole/wormhole.go:67]: Some comment on breaking API as above.

**Review by Jacalz — COMMENTED** [review]
I finally had time to have a short look through this. There are mostly a few suggestions on the creation of errors and some notes about braking API changes. Would it make sense to move more errors out to global variables?

**Comment by Jacalz** [comment]
I think it would be good if the changes weren't force-pushed every time. It makes it very hard to see changes between commits (if not impossible).

**Comment by piegamesde** [comment]
I had to rebase onto latest master in order to get some of the bug fixes for testing. For force pushes that are just amends and no rebases, you can click on the "force-pushed" in the GitHub UI to get the difference between both commits.
